File: js/src/focustree/focustree.ts
import { Writer } from "../structure"

export type FlagType = "None" | "Focus" | "Skip"

export interface FlagStat {
    focusCount: number
    skipCount: number
}

export interface Node {
    isLeaf(): boolean
    getFlag(): FlagType
    getChildren(): Node[]
    warning(message: string, stdout: Writer): void
}

export function extractSelectedLeaves(
    tree: Node,
    flagStat: FlagStat,
    stdout: Writer,
): Node[] {
    let focusedNodeArray = [] as Node[]
    findFocusedNodes(tree, focusedNodeArray, stdout)
    flagStat.focusCount = focusedNodeArray.length
    if (focusedNodeArray.length === 0) {
        focusedNodeArray.push(tree)
    }
    let leafArray = [] as Node[]
    focusedNodeArray.forEach((node) => {
        getLeaves(node, leafArray, flagStat)
    })
    return leafArray
}

function findFocusedNodes(
    node: Node,
    focusedNodeArray: Node[],
    stdout: Writer,
) {
    let flag = node.getFlag()
    if (flag == "Skip") {
        return
    }
    let initialLength = focusedNodeArray.length
    node.getChildren().forEach((child) => {
        findFocusedNodes(child, focusedNodeArray, stdout)
    })
    if (flag == "Focus") {
        if (focusedNodeArray.length > initialLength) {
            node.warning(
                "This node is marked as focused and it has focused descendents. The focus on this node will be ignored in favor of that of its descendents.",
                stdout,
            )
        } else {
            focusedNodeArray.push(node)
        }
    }
}

function getLeaves(node: Node, leafArray: Node[], flagStat: FlagStat) {
    if (node.getFlag() == "Skip") {
        flagStat.skipCount += 1
        return
    }
    node.getChildren().forEach((child) => {
        getLeaves(child, leafArray, flagStat)
    })
    if (node.isLeaf()) {
        leafArray.push(node)
    }
}

File: python/focustree/focustree.py
import enum
from typing import Protocol, TypeVar


class Flag(enum.Enum):
    NONE = 0
    FOCUS = 1
    SKIP = 2


TValue = TypeVar("TValue")


class Node(Protocol):
    def is_leaf(
        self,
    ) -> bool: ...
    def get_flag(
        self,
    ) -> Flag: ...
    def get_children(
        self,
    ) -> list["Node"]: ...
    def get_value(
        self,
    ) -> TValue: ...
    def warning(self, message: str): ...


def extract_selected_leaves(tree: Node) -> list[TValue]:
    focused_node_slice = []
    find_focused_nodes(tree, focused_node_slice)
    if len(focused_node_slice) == 0:
        focused_node_slice = [tree]
    leaf_slice = []
    for node in focused_node_slice:
        get_leaf_values(node, leaf_slice)
    return leaf_slice


def find_focused_nodes(node: Node, focused_node_slice: list[Node]):
    flag = node.get_flag()
    if flag == Flag.SKIP:
        return
    initial_length = len(focused_node_slice)
    for child in node.get_children():
        find_focused_nodes(child, focused_node_slice)
    if flag == Flag.FOCUS:
        if len(focused_node_slice) > initial_length:
            node.warning(
                "This node is marked as focused and it has focused descendents. "
                "The focus on this node will be ignored in favor of that of its descendents."
            )
        else:
            focused_node_slice.append(node)


def get_leaf_values(node: Node, leaf_slice: list[TValue]):
    if node.get_flag() == Flag.SKIP:
        return
    if node.is_leaf():
        leaf_slice.append(node.get_value())
    for child in node.get_children():
        get_leaf_values(child, leaf_slice)
